### PR TITLE
WT-10468 Enable WT Assertions in Release Mode

### PR DESCRIPTION
**Pull request by ajmorton**
Merge the PM-2897 project branch into develop. There are still some ongoing tasks (WT-10439, WT-10045, adding the `setParameter` to mongo) but these can or need to take place post-merge.

### CODE REVIEW

File: src/include/error.h [hunk 1]
```diff
@@ -124,67 +124,141 @@
...
+        size_t _offset;                                                                        \
+        if ((session) == NULL) {                                                               \
+            __wt_errx(                                                                         \
+              session, "A non-NULL session must be provided when unit testing assertions");    \
+            __wt_abort(session);                                                               \
+        }                                                                                      \
+                                                                                               \
+        /*                                                                                     \
+         * Normally these are two errx messages, but we need to stitch them both together into \
+         * unittest_assert_msg. To do so insert the second message immediately after the first \
```
Comment (clarissecheah) [inline src/include/error.h:157]: np: Add a fullstop at the end of this sentence.

File: bench/perf_run_py/perf_run.py [hunk 2]
```diff
@@ -337,3 +373,4 @@ def main():
 
 if __name__ == '__main__':
     main()
+
```
[the remark this answers is not in the document]
Comment (ajmorton) [reply bench/perf_run_py/perf_run.py]: Fixed in https://github.com/wiredtiger/wiredtiger/pull/8667/commits/5e90478e84e47cf776d3e84a6fce372b2f17396f 👍

File: src/include/error.h (hunk 1 above)
Comment (ajmorton) [reply src/include/error.h]: Fixed in https://github.com/wiredtiger/wiredtiger/pull/8667/commits/5e90478e84e47cf776d3e84a6fce372b2f17396f 👍

File: src/history/hs_rec.c [hunk 3]
```diff
@@ -145,40 +135,48 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
...
+                          type != WT_UPDATE_MODIFY &&
+                            (uint8_t)upd_type_full_diag != WT_UPDATE_MODIFY,
+                          "Duplicate modifies inserted into the history store can result in "
+                          "invalid reads");
+                    }
                 }
+                counter = hs_counter + 1;
             }
-            counter = hs_counter + 1;
+        } else {
```
Comment (clarissecheah) [inline src/history/hs_rec.c:176]: Is there a reason we left it as an if() inside and else() and didn't make this an else if (tw->start_ts == hs_start_ts)?

Comment (ajmorton) [reply src/history/hs_rec.c:176]: It was a minor cosmetic request in a previous PR: https://github.com/wiredtiger/wiredtiger/pull/8606#discussion_r1050263518. I'm happy to change it though as it's not a big thing.

Comment (clarissecheah) [reply src/history/hs_rec.c:176]: I remember now, thanks. Happy to leave it as is, was just curious.

**Review by clarissecheah — APPROVED** [review]
LGTM once all tests pass :)

**Pull request merged**